samza-core/src/main/java/org/apache/samza/operators/impl/SendToTableOperatorImpl.java

```diff
@@ -56,9 +55,9 @@ protected void handleInit(Context context) {
   @Override
   protected CompletionStage<Collection<Void>> handleMessageAsync(KV<K, V> message, MessageCollector collector,
       TaskCoordinator coordinator) {
-    table.put(message.getKey(), message.getValue());
     // there should be no further chained operators since this is a terminal operator.
-    return CompletableFuture.completedFuture(Collections.emptyList());
+    return table.putAsync(message.getKey(), message.getValue())
```

Comment:
with this change, the put order is not guaranteed any more.
Let's say a sequence of put event:
1. put(1, 10)
2. put (2, 20)
3. put(1, 5)
The 1st put fails at the 1st attempt and retry triggered.
2nd put succeed
3rd put succeed
then the retry for the 1st put succeeds.

Now the value for key 1 in the store is 10, while 5 is expected.

Reply:
Synced up offline about this. This doesn't change semantics in any ways than what it is today. 
The scenario outlined is only possible when task.max.concurrency is set to > 1. 
In case of task.max.concurrency, we expect the application to coordinate its shared states since there is no happens-before guarantee between messages delivered to the DAG.